r1cs/gadgets/std/src/groups/curves/short_weierstrass/short_weierstrass_jacobian.rs

```diff
@@ -345,22 +491,145 @@ for AffineGadget<P, ConstraintF, F>
+        let bits = bits.cloned().collect::<Vec<Boolean>>();
+
+        // Select any random T if self is infinity
+        let random_t = Self::alloc(cs.ns(|| "alloc random T"), || {
+            let mut rng = OsRng::default();
+            Ok(loop {
+                let r = SWProjective::<P>::rand(&mut rng);
+                if !r.is_zero() { break(r) }
+            })
+        })?;
+        let t = Self::conditionally_select(
+            cs.ns(|| "select self or random T"),
+            &self.infinity,
+            &random_t,
+            self
+        )?;
+
+        // Acc := [3] T = [2]*T + T
+        let mut acc = {
+            let mut t_copy = t.clone();
+            t_copy.double_in_place(cs.ns(|| "[2] * T"))?;
+            t_copy.add_unsafe(cs.ns(|| "[3] * T"), &t)
+        }?;
+
+        // Since len(bits) <= len(scalar field modulus), no arithemetic 
+        // edge case is met througout this loop. (For T = 0 we don't care.)    
+        for (i, bit) in bits.iter().enumerate()
+            // Skip the LSB (we handle it after the loop)
+            .skip(1)
+            // Scan over the scalar bits in big-endian order
+            .rev()
+            // Skip the MSB (already taken into account in the init of Acc)
+            .skip(1)
+        {
+            let mut cs = cs.ns(|| format!("bit {}", i));
+
+            // Q := k[i+1] ? T : −T
+            let neg_y = t.y.negate(cs.ns(|| "neg y"))?;
+            let selected_y = F::conditionally_select(
+                cs.ns(|| "select y or -y"),
+                bit,
+                &t.y,
+                &neg_y
+            )?;
+            let q = Self::new(t.x.clone(), selected_y, t.infinity);
+
+            // Acc := (Acc + Q) + Acc using double_and_add_internal
+            acc = acc.double_and_add_unsafe(cs.ns(|| "double and add"), &q)?;
+        }
+
+        // return (k[0] = 0) ? (Acc - T) : Acc
+        // Note that in this step both Acc and T are non-trivial and Acc != T, 
+        // but it might happen that Acc = -T in the single exceptional case of 
+        // exponent == scalar field modulus. 
+        // TODO: Let us think about whether asserting bits < scalar field modulus
+        // is appropriate. In this case, add_unsafe is secure.
+        let neg_t = t.negate(cs.ns(|| "neg T"))?;
+        let acc_minus_t = acc.add(
+            cs.ns(|| "Acc - T"),
+            &neg_t
+        )?;
+
+        let result = Self::conditionally_select(
+            cs.ns(|| "select acc or acc - T"),
+            &bits[0],
+            &acc,
+            &acc_minus_t
+        )?;
+
+        // If self was infinity, return 0 instead of result
+        let zero = Self::zero(cs.ns(|| "alloc 0"))?;
+        Self::conditionally_select(
+            cs.ns(|| "result or 0"),
+            &self.infinity,
+            &zero,
+            &result
+        )
+    }
+
     ///This will take [(4 + 1) * ceil(len(bits)/2)] constraints to put the x lookup constraint
```

Comment:
Let us elaborate on this comment. It is not clear to me what it asserts. 
Also, a web link to the implementation we refer to would be cool!